new Interface VCFIterator extends CloseableIterator<VariantContext>

lindenb opened the pull request:
### Description

As far as I know, there is no existing solution to handle a VCF file from an InputStream like System.in.
Here I introduce :
+ VCFIterator an `Iterator<VariantContext>` with the method `VCFheader getFileHeader()`
+ VCFIteratorBuilder a simple builder for the interface above (builds from uri,file and InputStream ). Decodes BCF,VCF, VCF.gz.


```
  VCFIterator r = new VCFIteratorBuilder().open(System.in);
  while (r.hasNext()) {
     System.out.println(r.next());
  }
  r.close();
```
```

### Checklist

- [X] Code compiles correctly
- [X] New tests covering changes and new functionality
- [X] All tests passing
- [X] Extended the README / documentation, if necessary
- [ ] Is not backward compatible (breaks binary or source compatibility)

magicDGS commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,210 @@
+     * @return the VCFIterator
+     * @throws OException
+     */
+    public VCFIterator open(final String uri) throws IOException {
```
I think that this method should use `ParsingUtils.openInputStream()` instead, to deal with `java.nio.Path`, `URL` and `File`.

magicDGS commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,210 @@
+     * @return the VCFIterator
+     * @throws IOException
+     */
+    public VCFIterator open(final File file) throws IOException {
```
I would love to have a `java.nio.Path` method too!

magicDGS replied:
For the `java.nio.Path` method, if the previous change is made, the simpler option will be:

```java
public VCFIterator open(final Path path) throws IOException {
    return open(new SeekablePathStream(path, null));
}
```

yfarjoun commented on src/test/java/htsjdk/variant/vcf/VCFIteratorTest.java:
```diff
@@ -0,0 +1,76 @@
+
+    @Test(dataProvider = "VcfFiles")
+    public void testUsingStreams(String uri, int nVariants) throws IOException {
+        VCFIterator r = new VCFIteratorBuilder().open(uri);
```
how is this test different from testUsingUri?

lindenb replied:
@yfarjoun ah you're right; I've changed this with a FileInputStream

yfarjoun commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,210 @@
+    public VCFIterator open(InputStream in) throws IOException {
+        if (in == null) {
+            throw new IllegalArgumentException("input stream is null");
+            }
```
remove spaces before `}`

yfarjoun commented on src/test/java/htsjdk/variant/vcf/VCFIteratorTest.java:
```diff
@@ -0,0 +1,80 @@
+        r.close();
+        in.close();
+    }
+
```
extra nl

yfarjoun commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,210 @@
+            this.inputStream.close();
+        }
+    }
+
```
extra nl

lindenb replied:
@yfarjoun back to you:

in Ioutils, I've added:
```

public static boolean isGZIPInputStream(BufferedInputStream in);
public static InputStream mayBeGZippedInputStream(InputStream in);
```

and I've added some tests for those two functions in IoUtilTest.

yfarjoun commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    /** implementation of VCFIterator, reading VCF */
+    private static class VCFReaderIterator extends AbstractIterator<VariantContext>implements VCFIterator {
```
space before `implements`

yfarjoun commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,210 @@
+        bufferedinput.reset();
+
+        //this is VCF or VCF.gz
+        if (bcfVersion == null) {
```
since you just tried reading the bcf header, coudl you flip the if statement and open the bcf if ! null?

yfarjoun commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,210 @@
+        if (bcfVersion == null) {
+            // try to read GZIP header
+            // see http://stackoverflow.com/questions/4818468/how-to-check-if-inputstream-is-gzipped
+            final byte[] signature = new byte[2];
```
Please refactor this useful piece of code to a function. Perhaps a static public function in BlockCompressedInputStream or IOUtil.

yfarjoun replied:
oh, and test it _please_.

magicDGS commented:
I still think that `java.nio.Path` support is a good adition to this PR, and it will be fairly simple. If not, I could open a PR after this one to add that support.

lindenb commented:
> , I could open a PR after this one to add that support.

@magicDGS yes please, I'm not familiar enough whith java.nio

magicDGS commented:
Perfect @lindenb. Could you please add a ticket after this is accepted (and maybe @yfarjoun could assign it to me) for that purpose? Just for tracking the work that should be done... Thanks in advance!

yfarjoun commented on src/main/java/htsjdk/samtools/util/IOUtil.java:
```diff
@@ -988,4 +988,45 @@ public static Path getPath(String uriString) throws IOException {
+        final BufferedInputStream bufferedinput;
+        /* it's already a buffered input stream, no need to wrap it again */
+        if(in instanceof BufferedInputStream) {
+            bufferedinput = BufferedInputStream.class.cast(in);
```
I think I'd avoid the premature optimization of trying to figure out if it's buffered "enough" and just wrap it in a buffer to be sure.

magicDGS replied:
I suppport @yfarjoun comment, and I will suggest to use the original input stream for the return, and use only the buffered one for the determination of the gzip signature...

lindenb replied:
@magicDGS I'm currently looking at this. Unless I'm misunderstanding there is no way to return the original InputStream because you need to consume the first bytes of the stream.

lindenb replied:
@magicDGS ah.. sorry for the duplicate: what @yfarjoun said below.

yfarjoun commented on src/main/java/htsjdk/samtools/util/IOUtil.java:
```diff
@@ -988,4 +988,45 @@ public static Path getPath(String uriString) throws IOException {
+        } else {
+            bufferedinput = new BufferedInputStream(in, 4);
+        }
+        if(isGZIPInputStream(bufferedinput)) {
```
space after if (and before parenthesis)

yfarjoun commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,200 @@
+        if (IOUtil.isUrl(uri)) {
+            try {
+                final URL url = new URL(uri); // this will throw if its not a
+                                              // url
```
no need for a single word in a newline

yfarjoun commented on src/test/java/htsjdk/samtools/util/IoUtilTest.java:
```diff
@@ -187,4 +191,54 @@ public void testFileTypeUnix(final String path, boolean expectedIsRegularFile) {
     }
+    
+    @Test
+    public void isGZIPInputStreamTest() throws IOException {
```
please use a dataprovider here too instead of testing two different scenarios in a single test.

magicDGS commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,200 @@
+     * @throws OException
+     */
+    public VCFIterator open(final String uri) throws IOException {
+        if (IOUtil.isUrl(uri)) {
```
Sorry, althought I said that I can add the `java.nio.Path` support, I realized that my change will change the behaviour of this method. To still support urls and files, but with the addition of `java.nio.Path` comming from a `String`, I think that here the best approach will be:

```java
 /**
  * creates a VCF iterator from an Path.
  * It detects if the stream is a BCF stream or a GZipped stream.
  * 
  * @param path a uri like string
  * @return the VCFIterator
  * @throws OException
  */
public VCFIterator open(final String path) throws IOException {
    return open(ParsingUtils.openInputStream(path, null));
}
```

This will simplify the code and do not change the behaviour in a later PR.

yfarjoun commented:
@magicDGS, that idea (wrapping and returning the unwrapped stream) seems "dangerous": If the underlying stream cannot reset back to the starting position (buffer is too small) it will not give you the beginning of the file anymore. Barring a way to detect how many byte can be read without irreversibly consuming data from the stream, you have to wrap it yourself.

Perhaps what we need is an override boolean (to mayBeGZIP) because the main use-case will be in the VCFIterator constructor and there we know that we have enough buffer that we do not want to re-wrap with another BufferedStream.

lindenb commented:
@magicDGS @yfarjoun @lbergelson thank you all for the reviews. Back to you. I've added @magicDGS 's suggestions for Path. (while I was here , I've closed some opened Streams in IoUtils.assertFilesEqual )

yfarjoun commented on src/main/java/htsjdk/samtools/util/IOUtil.java:
```diff
@@ -1009,21 +1011,15 @@ public static boolean isGZIPInputStream(final BufferedInputStream bufferedinput)
-            bufferedinput = new BufferedInputStream(in, 4);
-        }
-        if(isGZIPInputStream(bufferedinput)) {
+        final BufferedInputStream bufferedinput = new BufferedInputStream(in, 4);
```
I'm annoyed that we will have this extra layer around all our streamed inputs....

how about making a protected wrapper class around BufferedInputStream that is only used in `VCFIteratorBuilder` and then you can have a different signature function that doesn't wrap the Stream with another buffered Stream. Since this will be the main use-case, we will generally not have the extra layer.

ronlevine commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,200 @@
+import htsjdk.variant.variantcontext.VariantContext;
+
+/**
+ * An Class building {@link VCFReaderIterator}
```
Change `An` to `A`.

ronlevine commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,200 @@
+            }
+        // try to read the BCF header
+        final BufferedInputStream bufferedinput = new BufferedInputStream(in, 100);
+        bufferedinput.mark(100);
```
Can you use a static constant i.e. private static BCF_VERFSION_SIZE = 100 instead of a magic number? Can it be a smaller value?

ronlevine commented on src/main/java/htsjdk/samtools/util/IOUtil.java:
```diff
@@ -988,4 +991,38 @@ public static Path getPath(String uriString) throws IOException {
+     */
+    public static InputStream gunZipIfNeeded(final InputStream in) throws IOException {
+        /* BufferedInputStream is needed to decode the first bytes of the stream */
+        final BufferedInputStream bufferedinput = new BufferedInputStream(in, 4);
```
Why is this magic number 4 bytes? Can it be changed to a named static variable?

ronlevine commented on src/main/java/htsjdk/samtools/util/IOUtil.java:
```diff
@@ -988,4 +991,38 @@ public static Path getPath(String uriString) throws IOException {
+     */
+    public static boolean isGZIPInputStream(final BufferedInputStream bufferedinput) throws IOException {
+        // see http://stackoverflow.com/questions/4818468/how-to-check-if-inputstream-is-gzipped
+        final byte[] signature = new byte[2];
```
Can 2 be changed to static variable?

ronlevine commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,200 @@
+     * 
+     * @param path the path
+     * @return the VCFIterator
+     * @throws OException
```
Change to `IOException`. Tell when it will be thrown, same comment throughout file.

ronlevine commented on src/test/java/htsjdk/variant/vcf/VCFIteratorTest.java:
```diff
@@ -0,0 +1,90 @@
+                new Object[] { "src/test/resources/htsjdk/variant/serialization_test.bcf", 12 } };
+    }
+
+    private int countVariants(VCFIterator r) {
```
Make input variables final. Same comment for all methods in file.

ronlevine commented on src/test/java/htsjdk/variant/vcf/VCFIteratorTest.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    @Test(dataProvider = "VcfFiles")
+    public void testUsingUri(String uri, int nVariants) throws IOException {
```
Should `url` be `path`? Same comment got next test.

ronlevine commented on src/test/java/htsjdk/variant/vcf/VCFIteratorTest.java:
```diff
@@ -0,0 +1,90 @@
+    
+    @Test(dataProvider = "VcfFiles")
+    public void testUsingPath(String file, int nVariants) throws IOException {
+        Path path =Paths.get(file);
```
space after `=`.

ronlevine commented on src/test/java/htsjdk/variant/vcf/VCFIteratorTest.java:
```diff
@@ -0,0 +1,90 @@
+        Assert.assertNotNull(r.getFileHeader());
+        Assert.assertEquals(countVariants(r), nVariants);
+        r.close();
+        in.close();
```
I don't think this is needed, handled by `r.close()`.

ronlevine replied:
Can these 3 test be combined into one? Can there be round trip validation of the input file to ensure that what comes in is the same as what comes out?

ronlevine commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,200 @@
+        final BufferedInputStream bufferedinput = new BufferedInputStream(in, 100);
+        bufferedinput.mark(100);
+        final BCFVersion bcfVersion = BCFVersion.readBCFVersion(bufferedinput);
+        bufferedinput.reset();
```
Actually, this can be replaced by `BCFVersion#readBCFVersion(final InputStream stream)`.

lindenb commented:
@yfarjoun @lbergelson @ronlevine thank you for the reviews.

* I've fixed the typos, added `final`
* removed gunzipIfNeeded from IOUtils and moved it back to VCFIteratorBuilder
* added ` IOUtil.GZIP_SIGNATURE`
* using constants to create a bufferèsize for `new BufferedInputStream` in VCFIteratorBuilder
* simplified VCFIteratorBuilderTest

thanks

ronlevine commented on src/main/java/htsjdk/samtools/util/IOUtil.java:
```diff
@@ -988,4 +996,20 @@ public static Path getPath(String uriString) throws IOException {
     }
+    
+    /**
+     * Test whether a input stream looks like a GZIP input.
```
Add a blank line.

ronlevine commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,204 @@
+            //this is BCF
+            return new BCFInputStreamIterator(bufferedinput);
+        } else {
+            //
```
extra comment line

ronlevine commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,204 @@
+            throw new IllegalArgumentException("input stream is null");
+            }
+        // buffer must be large enough to contain the BCF header (+ min/max version) and/or GZIP signature
+        final int bufferSize = Math.max( BCFVersion.MAGIC_HEADER_START.length + 2*Byte.BYTES , IOUtil.GZIP_SIGNATURE.length);
```
extra space after `Math.max(`.

ronlevine commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,204 @@
+    }
+
+    /**
+     * creates a VCF iterator from a URI It detects if the stream is a BCF
```
Change URI to file path

droazen replied:
Why? `ParsingUtils.openInputStream()` handles more than just files (it handles http URIs, as well as URIs with schemes supported by NIO plugins)

ronlevine replied:
Correct. Ignore my comment.

ronlevine commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,204 @@
+            //this is VCF or VCF.gz, gunzip the input stream if needed
+            return new VCFReaderIterator( IOUtil.isGZIPInputStream(bufferedinput) ? 
+                    new GZIPInputStream(bufferedinput) : bufferedinput );
+        }
```
Should return null or throw an excpetion if `in` is not BCF, VCF or VCF.gz.

lindenb commented:
@yfarjoun @lbergelson @ronlevine thanks & back to you. Sorry for the late response, I do this PR on my 'workflow is running' time :-)

lindenb commented:
friendly ping for this one please.

yfarjoun commented:
@lbergelson could you take a look as well?

lbergelson commented on src/main/java/htsjdk/samtools/util/IOUtil.java:
```diff
@@ -97,6 +96,11 @@
+    /** signature of a gzip stream used by {@link isGZIPInputStream} 
+     * @see {@linkplain http://stackoverflow.com/questions/4818468/how-to-check-if-inputstream-is-gzipped}
+     */
+    public static final byte GZIP_SIGNATURE[]= {(byte) 0x1f, (byte) 0x8b};
```
It would be better I think to use the `GZIPInputStream.GZIP_MAGIC` constant directly instead of reproducing it here?

lbergelson commented on src/main/java/htsjdk/samtools/util/IOUtil.java:
```diff
@@ -97,6 +96,11 @@
 
     private static int compressionLevel = Defaults.COMPRESSION_LEVEL;
 
+    /** signature of a gzip stream used by {@link isGZIPInputStream} 
```
I think there are a few issues with the javadoc that my IDE pointed out to me.

 1.  to make that link work you need it to be `{@link #isGZIPInputStream}`
 2.  the `@see` doesn't need a `{@linkplain}`, it should be `* @see "http://stackoverflow.com/questions/4818468/how-to-check-if-inputstream-is-gzipped"` I believe.

lbergelson commented on src/main/java/htsjdk/samtools/util/IOUtil.java:
```diff
@@ -475,9 +479,13 @@ public static void assertFilesEqual(final File f1, final File f2) {
             while ((len1 = s1.read(buf1)) != -1) {
                 final int len2 = s2.read(buf2);
                 if (len1 != len2) {
+                    s1.close();
```
Good catch on these.  Instead of putting closes at all the exit points, I think it would make more sense to lift the creation of the FileInputStreams into a try-with-resources block and then they'll be guaranteed to be closed no matter where the function exits.

lbergelson commented on src/main/java/htsjdk/samtools/util/IOUtil.java:
```diff
@@ -988,4 +996,20 @@ public static Path getPath(String uriString) throws IOException {
+
+    /**
+     * Test whether a input stream looks like a GZIP input.
+     * @param bufferedinput the input stream. The buffer must be large enough to contain {@link GZIP_SIGNATURE}
```
similar javadoc issues to above

lbergelson commented on src/main/java/htsjdk/samtools/util/IOUtil.java:
```diff
@@ -988,4 +996,20 @@ public static Path getPath(String uriString) throws IOException {
+     * @throws IOException if `bufferedinput` cannot mark or reset the stream
+     * @see {@linkplain http://stackoverflow.com/questions/4818468/how-to-check-if-inputstream-is-gzipped}
+     */
+    public static boolean isGZIPInputStream(final BufferedInputStream bufferedinput) throws IOException {
```
Should this be just a simple `InputStream` but check markSupported and then throw if it isn't?  In any case it should probably explicitly check `markSupported()` and throw a helpful error message if it isn't.

lbergelson replied:
I noticed there's a similar method `SamStreams.isGzippedSAMFile`. Instead of just checking the header it actually tries to decompress the file and catches exceptions.  Your way may be slightly more elegant, but seems a little more brittle.  In either case, it should probably be refactored to use this method if that's possible.

lbergelson commented on src/main/java/htsjdk/variant/vcf/VCFIterator.java:
```diff
@@ -0,0 +1,56 @@
+ */
+public interface VCFIterator extends CloseableIterator<VariantContext> {
+    /** Returns the VCFHeader associated with this VCF/BCF file. */
+    public VCFHeader getFileHeader();
```
I would rename this to `getHeader()` since it's not necessarily read from a file.

lbergelson commented on src/main/java/htsjdk/variant/vcf/VCFIterator.java:
```diff
@@ -0,0 +1,56 @@
+import htsjdk.variant.variantcontext.VariantContext;
+
+/**
+ * An iterator used to read variants from an InputStream. It can be used to
```
This description seems misplaced, since the iterator itself doesn't have any of that functionality.  This comment seems like it should be on the `VCFIteratorBuilder()` instead.

lbergelson commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,207 @@
+ * @see {@link VCFIterator}
+ *
+ */
+
```
accidental linebreak here

lbergelson commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,207 @@
+        bufferedinput.reset();
+
+        try {
+            if (bcfVersion != null) {
```
what about gzipped bcf?

lindenb replied:
@lbergelson I need some advice on this point: I don't have a great experience with BCF. I've been working with `resources/test/htsjdk/variant/serialization_test.bcf `which I thought was a valid bcf file. But it's not compressed:

```
htsjdk$ find . -name "*.bcf" | xargs file
(...)
./build/resources/test/htsjdk/variant/serialization_test.bcf: data
```

on the other side, John Marshall says that a bcf is de-facto a bgz file   https://twitter.com/jomarnz/status/908768415971766272 .  It seems that there is no '*.bcf.gz' file ? Who's wrong ? Is there a true-valid bcf file under the 'resources/test' directory ?

Thanks

lindenb replied:
https://twitter.com/jomarnz/status/908775081236680705

> There is also “raw BCF” which bcftools outputs with -Ou. Historically HTSJDK has only supported raw BCF not real BCF

so, in htsjdk, the suffix '*.bcf' doesn't imply that the file is bgzipped ?

lbergelson replied:
So, it turns out I was confused.  I had assumed that it was always bgzipped, but it turns out htsjdk only handles uncompressed bcf at the moment. The spec recommends but doesn't require that bcf be compressed, but I think all real usage is with compressed files.  There's a ticket to fix it so that it handle proper block compressed bcfs https://github.com/samtools/htsjdk/issues/946.  No need for you to do anything I think.

lindenb replied:
@lbergelson ok thanks. I've worked on this, I'll push my latest commit tomorrow morning.

lbergelson commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,207 @@
+                return new BCFInputStreamIterator(bufferedinput);
+            } else {
+                //this is VCF or VCF.gz, gunzip the input stream if needed
+                return new VCFReaderIterator( IOUtil.isGZIPInputStream(bufferedinput) ? 
```
You're not special casing bgzip files here, but I assume that's fine since you're not allowing seeking?

lindenb replied:
>  since you're not allowing seeking?

true

lbergelson commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,207 @@
+     * @throws IOException
+     */
+    public VCFIterator open(final String path) throws IOException {
+        return open(ParsingUtils.openInputStream(path, null));
```
Could you add an overload that take the wrapper parameter as well?

lindenb replied:
I dont get that one: did you mean  adding an `open(String path,Function<SeekableByteChannel, SeekableByteChannel> wrapper)` ?

lbergelson replied:
Yes, that's exactly what I mean.

lbergelson commented on src/test/java/htsjdk/variant/vcf/VCFIteratorTest.java:
```diff
@@ -0,0 +1,87 @@
+
+    private void countVariants(final VCFIterator r, final int expectVariants) {
+        Assert.assertNotNull(r.getFileHeader());
+        final int nVariants =  (int) StreamSupport.stream(Spliterators.spliteratorUnknownSize(r, Spliterator.ORDERED), true).count();
```
As a `CloseableIterator`, `VCFIterator` already has this operation baked in as the inherited method `.stream()`

lbergelson commented on src/test/java/htsjdk/variant/vcf/VCFIteratorTest.java:
```diff
@@ -0,0 +1,87 @@
+        final int nVariants =  (int) StreamSupport.stream(Spliterators.spliteratorUnknownSize(r, Spliterator.ORDERED), true).count();
+        Assert.assertNotNull(r.getFileHeader());
+        Assert.assertEquals(nVariants, expectVariants);
+        r.close();
```
it's good practice to put a close like this in the finally of a try-finally block so it's executed even when the test fails assertions

lbergelson replied:
(although switching to use the baked in .stream() method closes it for you.)

lbergelson commented on src/test/java/htsjdk/samtools/util/IoUtilTest.java:
```diff
@@ -187,4 +191,35 @@ public void testFileTypeUnix(final String path, boolean expectedIsRegularFile) {
+        in.close();
+    }
+    @DataProvider(name = "mayBeGZippedInputStreamTestCases")
+    private Object[][] mayBeGZippedInputStreamTestCases() {
```
you added this dataprovider but you didn't actually do anything with it I think

lbergelson commented on src/test/java/htsjdk/samtools/util/IoUtilTest.java:
```diff
@@ -187,4 +191,35 @@ public void testFileTypeUnix(final String path, boolean expectedIsRegularFile) {
+    @DataProvider(name = "mayBeGZippedInputStreamTestCases")
+    private Object[][] mayBeGZippedInputStreamTestCases() {
+        return new Object[][]{
+                {"Hello World", Boolean.FALSE},
```
you can just use the constants `false` and `true` instead `Boolean.FALSE`,
also no need to put the hello world in the dataprovider if it's always the same.

lbergelson commented on src/test/java/htsjdk/variant/vcf/VCFIteratorTest.java:
```diff
@@ -0,0 +1,87 @@
+                new Object[] { "src/test/resources/htsjdk/variant/serialization_test.bcf", 12 } };
+    }
+
+    private void countVariants(final VCFIterator r, final int expectVariants) {
```
This is very nitpicky, but I would name this method something like `assertExpectedNumberOfVariants` since it performs an assertion.  I would expect a method called `countVariants` to return the number of variants, not to be a void method that sometimes blows up.

lindenb commented:
> @lindenb Sorry this sat for so long

@lbergelson no worries & thanks for the review !

lbergelson commented:
@lindenb This seems to have some compilation errors since the rebase.

lindenb commented:
@lbergelson yes, thanks ,  it saw it. I would like to have your feedback about the way htsjdk handles bcf. see https://github.com/samtools/htsjdk/pull/837#discussion_r139231337

lbergelson commented:
@lindenb I replied, turns out htsjdk bcf handling is worse than I thought it was, but you don't need to make any changes.

lindenb commented:
@lbergelson thanks &  back to you.

I hope I have made all the changes you wanted.  

* i'm now calling  IOUtil.isGZIPInputStream(stream)  moved and invoked from  SamStreams.isGzippedSAMFile() 
* I've created a 3rd implementation of VCFIterator in VCFIteratorBuilder that wraps a VCFFFileReader. This new simple class avoids to wrap the InputStream into the set of BufferedInputStream  of `open(InputStream)`.
* I've added some comments with links to this PR to explain why we only handle "vcf" / "vcf.gz" files at some points.
* I've added some tests that compress a VCF using gzip or bzip and  test if we can re-read the new compressed file.

cmnbroad commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,282 @@
+        // try to read a BCF header
+        bufferedinput.mark(sizeof_bcf_header);
+        final BCFVersion bcfVersion = BCFVersion.readBCFVersion(bufferedinput);
+        bufferedinput.reset();
```
I'd lobby to move this BCF-sniffing logic out into a public method somewhere, probably in BCFCodec; maybe to a static method that returns the BCFCodec or null on failure. In addition to keeping the knowledge of the BCF internals inside the BCFCodec, it would be immediately useful to solve [this problem](https://github.com/HadoopGenomics/Hadoop-BAM/issues/132) in Hadoop-BAM.

cmnbroad commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,282 @@
+     * using this class instead of VCFReaderIterator below,
+     * avoid to wrap an InputStream into a set of BufferedInputStream 
+     * */
+    private static class FileBasedVCFIterator
```
I'm not sure I understand the advantage of having this implementation class - why is this better than just having the open(File) method open a stream and delegate to open(InputStream). I THINK the input stream will get wrapped in (at least 1!) buffered stream either way.

cmnbroad commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,282 @@
+
+        VCFReaderIterator(final InputStream inputStream) {
+            this.inputStream = inputStream;
+            this.lineIterator = new AsciiLineReaderIterator(AsciiLineReader.from(inputStream));
```
It would be better to delegate to the codec's makeSourceFromStream method here.

cmnbroad commented on src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java:
```diff
@@ -0,0 +1,282 @@
+        private final VCFHeader vcfHeader;
+
+        BCFInputStreamIterator(final InputStream inputStream) {
+            this.inputStream = new PositionalBufferedStream(inputStream);
```
Same here; delegate to the codec 's makeSourceFromStream method.

lbergelson commented on src/main/java/htsjdk/samtools/SamStreams.java:
```diff
@@ -66,26 +67,11 @@ public static boolean isBAMFile(final InputStream stream)
      * is and false otherwise.
+     * @see {@link IOUtil#isGZIPInputStream(InputStream)}
      */
     public static boolean isGzippedSAMFile(final InputStream stream) {
```
👍

lindenb commented:
@lbergelson @cmnbroad thank you for the reviews !

- I've deprecated isGzippedSAMFile
- created BCF2Codec.tryReadBCFVersion
- fixed the typo
- deleted FileBasedVCFIterator
- used the codec.makeSourceFromStream
- etc..

lindenb commented:
friendly ping for this one :-)

romanzenka commented:
Hello, I could really use this code. What can I do to make it happen?

lindenb commented:
I gave up for this PR , I'm now using my private src : https://github.com/lindenb/jvarkit/blob/master/src/main/java/com/github/lindenb/jvarkit/util/vcf/VcfIteratorImpl.java & https://github.com/lindenb/jvarkit/blob/master/src/main/java/com/github/lindenb/jvarkit/util/vcf/VcfIterator.java

romanzenka commented:
@lindenb I can totally use your local fork for a while - but maybe if you could let me know what is the holdup here, I could try to "carry the torch"?
Is it fair that what is left are @cmnbroad and @lbergelson change requests, and then you need to figure out that weird "coverage changes not visible in diff" codecov error?

lindenb commented:
>  lindenb:vcfiter 

https://github.com/lindenb/htsjdk/tree/vcfiter

lbergelson commented:
Ack.. this was one of those totally reasonable and helpful pr's that get dogpiled on by every reviewer in the world and then we all just gave up on...  Lets fix that...

romanzenka commented:
I made PR #1245, will continue my work there!

Many thanks to @lindenb for getting the implementation 99% there!

lbergelson commented:
@lindenb Fair enough... sorry it became such a mess.  Thanks for doing this work.  Should we close this then?

lindenb commented:
closing, moved to https://github.com/samtools/htsjdk/pull/1245
